tools/pinning/common/export-pinned-dependencies.sh

```diff
@@ -40,5 +40,5 @@ fi
 poetry lock >&2
 trap 'rm poetry.lock' EXIT
 
-# We need to remove local packages from the output.
-poetry export --without-hashes | sed '/^acme @/d; /certbot/d;'
+# We need to remove local packages and extras specifiers from the output
+poetry export --without-hashes | sed '/^acme @/d; /certbot/d; s/\[[^][]*\]//g;'
```

Comment:
This regex turns `docker[ssh]` into `docker`, which keeps our `requirements.txt` file the same as what it was before. It is unclear to me why `poetry export` started spitting out extras all of a sudden.

I am not sure whether this is the right approach. That we are using the output of `poetry export` as constraints for `pip` is perhaps not how `poetry export` is meant to be used. With Brad and Adrien unavailable, this is what I could come up with to get things working again.

Reply:
So what `poetry` imagines is that you'll use their whole system and not need to touch `pip` at all, though it does look like they're [using it under the hood](https://github.com/python-poetry/poetry/blob/c8ae868472d9e6e4fc281adc08f2d14e374d35c8/src/poetry/installation/installer.py#L555). By using `export` we're saying "do this much and no more, give us what you have". So I suspect the fact we're using pip is not in and of itself a problem.

As for the extras, the docs seem to imply that you have to [specifically ask for extras](https://python-poetry.org/docs/pyproject/#extras), so that's strange. Ah, it might be [this bug](https://github.com/python-poetry/poetry/commit/fdec38b1e03f8072e6df57f6376a85abb7803fec) that was recently fixed? Does updating to `1.2.0b3` fix it? Ideally I'd like to not have to use regexes to customize things, given that the whole point of poetry is to be a modern configurable system.

Reply:
>Does updating to 1.2.0b3 fix it?

No, using the latest Poetry still includes the extras.

>imply that you have to [specifically ask for extras](https://python-poetry.org/docs/pyproject/#extras)

I don't follow, that part of the Poetry documentation confuses me. I don't know what that `-E` flag to `poetry export` is actually supposed to do.

According to `pip` documentation, requirements files can list extras. [Constraints files cannot](https://pip.pypa.io/en/latest/user_guide/#constraints-files).

`docker-compose` depends on `docker[ssh]`, which resolves to `docker`+`paramiko`. Listing `docker[ssh]` in a requirements file is a valid thing to do, according to the spec. 

I think Poetry is correct in producing that output. It just doesn't suit our use case of using it as a constraints file, which should just be a flattened list of dependencies and versions.

Do you think I am landing in the right place with this understanding? I'm not sure where else to go with it.

Reply:
Yeah, I also find it confusing, that was mostly a hopeful guess. Probably they meant scoped extras of your project, not of its dependencies.

My main worry here is accidentally removing something that turns out to be necessary to specify. I see that `paramiko` is in there, but how did it get there? If poetry already expanded `docker[ssh]`, why leave it?

Reply:
>My main worry here is accidentally removing something that turns out to be necessary to specify

Since extras are illegal in constraints files, this feels somewhat safe to me. I am not aware of any other ways that square brackets could appear, having read the [requirements file format](https://pip.pypa.io/en/latest/reference/requirements-file-format/#using-environment-variables) and the [PEP 508 grammar for requirement specifiers](https://peps.python.org/pep-0508/#grammar).

>If poetry already expanded docker[ssh], why leave it?

I don't know, sorry. 

Would you prefer to wait for more input from others rather than merging this? I don't think we urgently need to repin anything, this was originally triggered by the pyjwt dependabot alert.

Reply:
If this is the only option we can find, I'd be ok with it, but it bothers me conceptually. The whole point of using poetry is to not having to do manual finagling ourselves, and to use a robust system to manage dependencies. I'll take another look at poetry to see if there's an idiomatic way to handle this; if there's not, we could maybe even ask the maintainers to include an option to not export it with extras specifiers/ submit a PR there.

Reply:
I get the behavior of having `docker[ssh]` included too if I upgrade poetry and/or its dependencies before running the repin script on `master`. If I create a fresh virtual environment on `master` and use the version of poetry currently pinned there, I do not get `docker[ssh]` in the requirements file.

After that run of the repin script though, the newly pinned version of poetry & friends **will** include `docker[ssh]` on future runs which is a problem. I didn't track down exactly what changed to cause this, but my vote is we currently pin back poetry and/or its dependencies in `tools/pinning/current/pyproject.toml` while we/I investigate this.

What do you two think?

Reply:
I took the approach I described in my last post at https://github.com/certbot/certbot/pull/9366 if either of you want it, but thinking about this a bit more, I also suspect Alex's approach here will work just fine and may be our best option until/unless poetry changes things again.

Reply:
Why did you close that PR? If we think they'll change it back then yeah pinning sounds great. This might be the sort of problem that takes communication to solve.

Reply:
I am happy to pin Poetry in the short term. We will no doubt want to upgrade Poetry again soon (when [support for yanked releases ships](https://github.com/python-poetry/poetry/pull/5841)) so that we can make sure that we don't accidentally introduce yanked releases into our automatic repinning (https://github.com/certbot/certbot/issues/9336).

I'll grab the commit, thank you!

Reply:
> Why did you close that PR?

My PR isn't going to work on its own due to things like the pylint issues fixed in this PR. My understanding is Alex is planning to pull that commit in here and after dependencies are regenerated, I think this should be (pretty close to) mergeable.